scripts/upgrade-facets.js

```diff
@@ -0,0 +1,244 @@
+
+  if (!protocolAddress) {
+    return addressNotFound("ProtocolDiamond");
+  }
+
+  if (!accessControllerAddress) {
+    return addressNotFound("AccessController");
+  }
+
+  // Get AccessController abstraction
+  const accessController = await ethers.getContractAt("AccessController", accessControllerAddress);
+
+  // Check that caller has upgrader role.
+  const hasRole = await accessController.hasRole(Role.UPGRADER, admin);
+  if (!hasRole) {
+    console.log("Admin address does not have UPGRADER role");
+    process.exit(1);
+  }
+
+  // Deploy new facets
+  const deployedFacets = await deployProtocolHandlerFacets(
+    protocolAddress,
+    Facets.addOrUpgrade,
+    maxPriorityFeePerGas,
+    false
+  );
+
+  // Cast Diamond to DiamondCutFacet and DiamondLoupeFacet
+  const diamondCutFacet = await ethers.getContractAt("DiamondCutFacet", protocolAddress);
+  const diamondLoupe = await ethers.getContractAt("DiamondLoupeFacet", protocolAddress);
+
+  // All handler facets currently have no-arg initializers
+  let initFunction = "initialize()";
+  let initInterface = new ethers.utils.Interface([`function ${initFunction}`]);
+  let callData = initInterface.encodeFunctionData("initialize");
+
+  // manage new or upgraded facets
+  for (const newFacet of deployedFacets) {
+    console.log(`\n📋 Facet: ${newFacet.name}`);
+
+    // Get currently registered selectors
+    const oldFacet = contracts.find((i) => i.name === newFacet.name);
+    let registeredSelectors;
+    if (oldFacet) {
+      // Facet already exists and is only upgraded
+      registeredSelectors = await diamondLoupe.facetFunctionSelectors(oldFacet.address);
+    } else {
+      // Facet is new
+      registeredSelectors = [];
+    }
+
+    // Remove old entry from contracts
+    contracts = contracts.filter((i) => i.name !== newFacet.name);
+    deploymentComplete(newFacet.name, newFacet.contract.address, [], contracts);
+
+    // Get new selectors from compiled contract
+    const selectors = getSelectors(newFacet.contract, true);
+    const newSelectors = selectors.selectors.remove([initFunction]);
+
+    // Determine actions to be made
+    let selectorsToReplace = registeredSelectors.filter((value) => newSelectors.includes(value)); // intersection of old and new selectors
+    let selectorsToRemove = registeredSelectors.filter((value) => !selectorsToReplace.includes(value)); // unique old selectors
+    let selectorsToAdd = newSelectors.filter((value) => !selectorsToReplace.includes(value)); // unique new selectors
+
+    // Skip selectors if set in config
+    const selectorsToSkip = Facets.skip[newFacet.name] ? Facets.skip[newFacet.name] : [];
+    selectorsToReplace = removeSelectors(selectorsToReplace, selectorsToSkip);
+    selectorsToRemove = removeSelectors(selectorsToRemove, selectorsToSkip);
+    selectorsToAdd = removeSelectors(selectorsToAdd, selectorsToSkip);
+
+    // Logs
```

Comment:
I would move this logging in this section to after the actual work is done, otherwise if that fails, then this output might be confusing.

Reply:
Moved.